File: practices/chapter4-queue/ch_4_queue.js
class Queue {
  constructor() {
    this.queue = []
  }

  enqueue(data) { // add data to queue
    this.queue.unshift(data) // from start
    return this
  }

  dequeue() { // read data from queue
    if(this.size()) {
      return this.queue.pop() // from end
    }
    return 'this queue is empty now.'
  }

  size() {
    return this.queue.length
  }
}

function enqueueSample() {
  const queue = new Queue()
  const size = queue.enqueue('a').enqueue('b').enqueue('c').size()
  console.log(size)
}

function dequeueSample() {
  const queue = new Queue()
  queue.enqueue('a').enqueue('b').enqueue('c')
  while (queue.size()) {
    console.log(queue.dequeue())
  }
}

enqueueSample()
dequeueSample()

File: practices/chapter4-queue/ch_4_queue.py
class Queue():
  def __init__(self):
    self.queue = []

  def enqueue(self, data): # 插入資料
    self.queue.insert(0, data)
    return self

  def dequeue(self): # 讀取資料(讀完拔除)
    if self.size():
      return self.queue.pop()
    return "it's empty queue."

  def size(self):
    return len(self.queue)

def enqueue_sample():
  queue = Queue()
  queue.enqueue(1).enqueue(2).enqueue(3)
  print('lenght of queue is: ', queue.size())

def dequeue_sample():
  queue = Queue()
  queue.enqueue(1).enqueue(2).enqueue(3)
  while queue.size() > 0:
    print('read queue: ', queue.dequeue())

if __name__ == "__main__":
    enqueue_sample()
    dequeue_sample()